File: Chapter5/5.9/palindrome/java/palin.java
import java.util.Scanner;

public class palin {
    public static boolean isPalin(int x){
	if (x<0) return false;
	if (x==0) return true;
	int denom = (int)Math.pow(10,(int)(Math.log10(x)));
	while (x>9) {
	    if (x/denom != x%10) return false;
	    x %= denom;
	    x /= 10;
	    denom /= 100;
	}
	return true;
    }

    public static void main(String[] args){
	Scanner in = new Scanner(System.in);
	int x = in.nextInt();
	if (isPalin(x)){
	    System.out.println("true");
	}else System.out.println("false");
    }
}

File: Chapter5/5.9/palindrome/cpp/palin.cpp
#include <iostream>
#include <cmath>

bool isPalin(int x){
  if (x<0) return false;
  if (x==0) return true;

  int denom = pow(10,static_cast<int>(log10(x))) ;
  while (x>9) {
    int leftDigit = x/denom;
    int rightDigit = x%10;
    using namespace std;
    cout<<leftDigit<<":"<<rightDigit<<endl;
    if (leftDigit != rightDigit) return false;
    x %= denom;
    x /= 10;
    cout<<"x:"<<x<<endl;
    denom /= 100;   
  }
  return true;
}

int main(){
  using namespace std;
  while (1) {
    int x;
    cin>>x;
    if (x==-1) return 0;
    if (isPalin(x)){ 
      cout<<"true"<<endl;
    }else{
      cout<<"false"<<endl;
    }    
  }
}
